src/client/Microsoft.Identity.Client/TelemetryCore/TelemetryClientExtensions.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.IdentityModel.Abstractions;
+
+namespace Microsoft.Identity.Client.TelemetryCore
+{
+    /// <summary>
+    /// Extension methods used for telemetry client.
+    /// </summary>
+    internal static class TelemetryClientExtensions
+    {
+        /// <summary>
+        /// Gets a set of <see cref="ITelemetryClient"/>s which are enabled for a given <paramref name="eventName"/>.
+        /// </summary>
+        /// <param name="clients">Set of clients to check.</param>
+        /// <param name="eventName">Event name to evaluate.</param>
+        /// <returns>The relevant set of <see cref="ITelemetryClient"/>s.</returns>
+        internal static IEnumerable<ITelemetryClient> GetEnabledClients(this IEnumerable<ITelemetryClient> clients, string eventName)
+        {
+            return clients?.Where(c => c.IsEnabled(eventName));
+        }
+
+        /// <summary>
+        /// Sends the same input events to each telemetry client.
+        /// </summary>
+        /// <param name="clients">Clients to emit telemetry to.</param>
+        /// <param name="eventDetails">Telemetry details to emit.</param>
+        internal static void TrackEvent(this IEnumerable<ITelemetryClient> clients, TelemetryEventDetails eventDetails, string eventName)
+        {
+            foreach (var client in clients.GetEnabledClients(eventName))
```

Comment:
>clients.GetEnabledClients(eventName) [](http://example.com/codeflow?start=35&length=36)

If you collapse the enabled check into this method, then you'll always execute the work to build out the event details no matter if any clients have that event enabled.

Better to separately check if any client is enabled for the event THEN create the details and pass it to the enabled clients.

Reply:
I am building the event details separately and not when logging. The idea is to have the event details populated as we go through the request and in future when we add the error details or L1/L2 cache details we might have to pass the event details object around. See [this](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/nebharg%2FLogTelemetry/src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs#L123)